### PR TITLE
Move tests from sandbox to canton

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Implements `withCantonSandbox`, which directly replaces `withSandbox` in the various tests.
Also implements a haskell FreePort to support this correctly. Should cooperate directly with the existing scala implementation

TODO: verify co-operability

### CODE REVIEW

File: compiler/damlc/tests/src/DA/Test/Repl.hs [hunk 1]
```diff
@@ -118,7 +117,7 @@ testConnection damlc scriptDar ledgerPort mbTokenFile mbCaCrt = do
         [ "alice <- allocatePartyWithHint \"Alice\" (PartyIdHint \"Alice\")"
         , "debug alice"
         ]
-    let regexString = "^.*daml>.*: 'Alice'\ndaml> Goodbye.\n$" :: String
+    let regexString = "^.*daml>.*: 'Alice[^\n]*\ndaml> Goodbye.\n$" :: String
```
Comment (remyhaemmerle-da) [inline compiler/damlc/tests/src/DA/Test/Repl.hs:120]:
```suggestion
    let regexString = "^.*daml>.*: 'Alice:[a-f0-9]+\ndaml> Goodbye.\n$" :: String
```
This does not work ?

File: compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs [hunk 2]
```diff
@@ -278,18 +276,18 @@ functionalTests replClient replLogger serviceOut options ideState = describe "re
           , input "bob <- allocatePartyWithHint \"Bob\" (PartyIdHint \"bob\")"
           , input "proposal <- pure (T alice bob)"
           , input "debug proposal.proposer"
-          , matchServiceOutput "'alice'"
+          , matchServiceOutput "'alice:"
           , input "debug proposal.accepter"
-          , matchServiceOutput "'bob'"
+          , matchServiceOutput "'bob:"
```
Comment (remyhaemmerle-da) [inline compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs:281]: can we use similar regexp as above?

File: compiler/damlc/tests/BUILD.bazel [hunk 3]
```diff
@@ -689,8 +689,8 @@ da_haskell_test(
         "//compiler/damlc/stable-packages",
         "//compiler/repl-service/server:repl_service_jar",
         "//daml-script/daml:daml-script.dar",
-        "//ledger/sandbox-on-x:app",
-    ],
+        "@canton//:lib",
+    ] + (["@sysctl_nix//:bin/sysctl"] if is_darwin else []),
```
Comment (remyhaemmerle-da) [inline compiler/damlc/tests/BUILD.bazel:693]: naive question, why do we need this  ?

**Review by remyhaemmerle-da — APPROVED** [review]
Thanks !

File: compiler/damlc/tests/src/DA/Test/Repl.hs (hunk 1 above)
Comment (samuel-williams-da) [reply compiler/damlc/tests/src/DA/Test/Repl.hs:120]: I'd have to double check, for example I think there would be a close `'` added at least. I'll check

File: compiler/damlc/tests/BUILD.bazel (hunk 3 above)
Comment (samuel-williams-da) [reply compiler/damlc/tests/BUILD.bazel:693]: For FreePort, in order to get the dynamic port range, we need `sysctl` on MacOS.
This is the same for FreePort in scala :)

**Pull request merged**